File: craftinteractive/craft_interactive/report/machine_maintenance_report/machine_maintenance_report.py
import frappe

def execute(filters=None):
    filters = filters or {}
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    if filters.get("consolidated"):
        return [
            {"label": "Machine", "fieldname": "machine_name", "fieldtype": "Link", "options": "Item", "width": 200},
            {"label": "Total Cost", "fieldname": "cost", "fieldtype": "Currency", "width": 150},
        ]

    return [
        {"label": "Machine", "fieldname": "machine_name", "fieldtype": "Link", "options": "Item", "width": 200},
        {"label": "Maintenance Date", "fieldname": "maintenance_date", "fieldtype": "Date", "width": 200},
        {"label": "Technician", "fieldname": "technician", "fieldtype": "Link", "options": "Employee", "width": 200},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 200},
        {"label": "Cost", "fieldname": "cost", "fieldtype": "Currency", "width": 200},
        {"label": "Row Color", "fieldname": "color", "fieldtype": "Data", "hidden": 1},
    ]


def get_data(filters):

    conditions = "" 

    if filters.get("machine_name"):
        conditions += " AND machine_name = %(machine_name)s"

    # Technician filter
    if filters.get("technician"):
        conditions += " AND technician = %(technician)s"

    # DateRange filter
    if filters.get("from_date"):
        start_date, end_date = filters.get("from_date")
        
        filters["start_date"] = start_date
        filters["end_date"] = end_date

        conditions += " AND maintenance_date BETWEEN %(start_date)s AND %(end_date)s"


    records = frappe.db.sql("""
    SELECT 
        name, machine_name, maintenance_date, technician, status, cost
    FROM `tabMachine Maintenance`
    WHERE docstatus < 2 {conditions}
    ORDER BY maintenance_date DESC
""".format(conditions=conditions), filters, as_dict=True)

    if filters.get("consolidated"):
        return consolidate(records)

    return highlight(records)


def consolidate(records):
    summary = {}

    for d in records:
        summary.setdefault(d.machine_name, 0)
        summary[d.machine_name] += d.cost or 0

    result = []
    for machine, total in summary.items():
        result.append({
            "machine_name": machine,
            "cost": total
        })

    return result



def highlight(records):
    for d in records:
        if d.status == "Overdue":
            d["color"] = "red"
        elif d.status == "Scheduled":
            d["color"] = "yellow"
        elif d.status == "Completed":
            d["color"] = "green"
        else:
            d["color"] = ""
    return records

File: craftinteractive/craft_interactive/report/machine_maintenance_report/machine_maintenance_report.js
// Copyright (c) 2025, Balaji B and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Machine Maintenance Report"] = {
	formatter(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data || !data.status) {
            return value;
        }

        let color = "";

        if (data.status === "Overdue") {
            color = "background-color:#ffcccc !important;";   // light red
        }
        else if (data.status === "Scheduled") {
            color = "background-color:#fff2b3 !important;";   // light yellow
        }
        else if (data.status === "Completed") {
            color = "background-color:#d6f5d6 !important;";   // light green
        }

        return `<span style="${color} display:block; width:100%; height:100%;">${value}</span>`;
    },
	"filters": [
		{
        "fieldname": "machine_name",
        "label": "Machine",
        "fieldtype": "Link",
        "options": "Item"
    },
    {
        "fieldname": "technician",
        "label": "Technician",
        "fieldtype": "Link",
        "options": "Employee"
    },
    {
        "fieldname": "from_date",
        "label": "From Date",
        "fieldtype": "DateRange"
    },
    {
        "fieldname": "consolidated",
        "label": "Consolidated",
        "fieldtype": "Check",
        "default": 0
    }
	]
};
